### PR TITLE
[TRAFODION-2468] DDL HA , step 2 changes(please refer jira).

### PR DESCRIPTION
**Pull request by prashanth-vasudev**

### CODE REVIEW

**Comment by DaveBirdsall** [comment]
@sbroeder, @shangshengtung: Could you review this?

**Comment by shangshengtung** [comment]
Will have a review session to walk through the change next week.

File: core/sqf/src/seatrans/tm/hbasetmlib2/src/main/java/org/trafodion/dtm/HBaseTxClient.java [hunk 1]
```diff
@@ -1232,8 +1203,218 @@ public void run() {
...
+                                " and tolerating UnknownTransactionExceptions");
+                    txnManager.doCommit(ts, true /*ignore UnknownTransactionException*/);
+                    if(useTlog && useForgotten) {
+                        long nextAsn = tLog.getNextAuditSeqNum((int)TransactionState.getNodeId(txID));
+                        tLog.putSingleRecord(txID, ts.getCommitId(), "FORGOTTEN", null, forceForgotten, nextAsn);
+                    }
+                } else if (ts.getStatus().equals(TransState.STATE_ABORTED.toString())) {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("TRAF RCOV THREAD:Redriving abort for " + txID);
+                    txnManager.abort(ts);
```
Comment (prashanth-vasudev) [inline core/sqf/src/seatrans/tm/hbasetmlib2/src/main/java/org/trafodion/dtm/HBaseTxClient.java]: @sbroeder :  Do we need true /*ignore UnknownTransactionException*/ for abort as well?

File: core/sqf/src/seatrans/tm/hbasetmlib2/src/main/java/org/trafodion/dtm/HBaseTxClient.java [hunk 2]
```diff
@@ -1232,8 +1203,218 @@ public void run() {
...
+                }
+  
+            } catch (UnsuccessfulDDLException ddle) {
+                LOG.error("UnsuccessfulDDLException encountered by Recovery Thread. Registering for retry. txID: " + txID + "Exception " , ddle);
+  
+                //Note that there may not be anymore redrive triggers from region server point of view for DDL operation.
+                //Register this DDL transaction for subsequent redrive from Audit Control Event.
+                //TODO: Launch a new Redrive Thread out of auditControlPoint().
+                TmDDL tmDDL = hbtx.getTmDDL();
+                tmDDL.setState(txID,"REDRIVE");
```
Comment (prashanth-vasudev) [inline core/sqf/src/seatrans/tm/hbasetmlib2/src/main/java/org/trafodion/dtm/HBaseTxClient.java]: @shangshengtung :  There needs to be clear distinction of responsibility between worker thread and recovery thread. Here there is not need to change the state of txID in tmDDL. Recovery thread can retry in subsequent loop.

File: core/sqf/src/seatrans/tm/hbasetmlib2/src/main/java/org/trafodion/dtm/HBaseTxClient.java (hunk 1 above)
Comment (prashanth-vasudev) [inline core/sqf/src/seatrans/tm/hbasetmlib2/src/main/java/org/trafodion/dtm/HBaseTxClient.java]: abort does not take ignore unknownTransactionException flag. We can add if abort starts to support this flag.

File: core/sqf/src/seatrans/tm/hbasetmlib2/src/main/java/org/trafodion/dtm/HBaseTxClient.java (hunk 2 above)
Comment (prashanth-vasudev) [inline core/sqf/src/seatrans/tm/hbasetmlib2/src/main/java/org/trafodion/dtm/HBaseTxClient.java]: will remove the REDRIVE tag tmDDL and set recovery thread to loop back for recheck.

File: core/sqf/src/seatrans/tm/hbasetmlib2/src/main/java/org/trafodion/dtm/HBaseTxClient.java (hunk 1 above)
Comment (sbroeder) [inline core/sqf/src/seatrans/tm/hbasetmlib2/src/main/java/org/trafodion/dtm/HBaseTxClient.java]: OK.  Sounds good.

**Comment by shangshengtung** [comment]
Looks good.

**Pull request merged**